sql/src/main/java/io/crate/operation/projectors/FetchProjector.java

```diff
@@ -21,384 +21,426 @@
 
 package io.crate.operation.projectors;
 
-import com.carrotsearch.hppc.IntArrayList;
+import com.carrotsearch.hppc.IntContainer;
 import com.carrotsearch.hppc.IntObjectOpenHashMap;
-import com.carrotsearch.hppc.LongArrayList;
-import com.google.common.collect.Sets;
-import com.google.common.util.concurrent.MoreExecutors;
-import com.google.common.util.concurrent.SettableFuture;
+import com.carrotsearch.hppc.IntSet;
+import com.carrotsearch.hppc.cursors.IntCursor;
+import com.carrotsearch.hppc.cursors.IntObjectCursor;
+import com.google.common.collect.Iterables;
+import io.crate.Streamer;
+import io.crate.core.collections.Bucket;
 import io.crate.core.collections.Row;
-import io.crate.core.collections.RowN;
-import io.crate.executor.transport.*;
+import io.crate.executor.transport.NodeFetchRequest;
+import io.crate.executor.transport.NodeFetchResponse;
+import io.crate.executor.transport.StreamBucket;
+import io.crate.executor.transport.TransportFetchNodeAction;
 import io.crate.jobs.ExecutionState;
 import io.crate.metadata.Functions;
 import io.crate.metadata.PartitionName;
-import io.crate.metadata.ReferenceInfo;
+import io.crate.metadata.TableIdent;
 import io.crate.operation.Input;
 import io.crate.operation.InputRow;
-import io.crate.operation.collect.CollectExpression;
-import io.crate.operation.fetch.PositionalBucketMerger;
-import io.crate.operation.fetch.PositionalRowDelegate;
-import io.crate.operation.fetch.RowInputSymbolVisitor;
-import io.crate.planner.symbol.Reference;
+import io.crate.operation.fetch.FetchRowInputSymbolVisitor;
+import io.crate.planner.node.fetch.FetchSource;
 import io.crate.planner.symbol.Symbol;
+import io.crate.planner.symbol.Symbols;
 import org.apache.lucene.util.BytesRef;
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.common.logging.ESLogger;
 import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.threadpool.ThreadPool;
 
-import javax.annotation.Nullable;
 import java.util.*;
-import java.util.concurrent.CancellationException;
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.Executor;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 
 public class FetchProjector extends AbstractProjector {
 
-    private PositionalBucketMerger downstream;
+    enum Stage {
+        INIT,
+        COLLECT,
+        FETCH,
+        FINALIZE
+    }
+
+    private AtomicReference<Stage> stage = new AtomicReference<>(Stage.INIT);
+    private final Object failureLock = new Object();
+
+    private final AtomicReference<Throwable> failure = new AtomicReference<>();
+    private final FetchRowInputSymbolVisitor.Context collectRowContext;
+    private RowReceiver downstream;
```

Comment:
is there a reason that the FetchProjector uses its own downstream attribute instead of using the one from the base class?